### PR TITLE
feat: accelerate all-pairs layout with grid data structure

### PR DESCRIPTION
**Pull request by k-qy**
_Not yet ready for review_

Currently implementing this: [hashing.pdf](https://github.com/penrose/penrose/files/6657066/hashing.pdf)

Repro: `roger watch graph-domain/big-graph.sub graph-domain/disjoint-rects-big.sty graph-domain/graph-theory.dsl`

Related issues: #596 

---

# Description

Related issue/PR: # (issue/PR)

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

# Implementation strategy and design decisions

Include a high-level summary of the implementation strategy and list important design decisions made, if any.

# Examples with steps to reproduce them

# Checklist

- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass locally using `yarn test`
- [ ] I ran `yarn docs` and there were no errors when generating the HTML site
- [ ] My code follows the style guidelines of this project (e.g.: no ESLint warnings)

# Open questions

Questions that require more discussion or to be addressed in future development:

### CODE REVIEW

File: packages/core/src/engine/Autodiff.ts [hunk 1]
```diff
@@ -1125,857 +1127,861 @@ export const fns = {
...
+                log.trace("args", childNames);
+                throw Error("expected three args to if cond");
+            }
+
+            stmt = `const ${parName} = ${childNames[0]} ? ${childNames[1]} : ${childNames[2]};`;
+        } else if (op === "+ list") {
+            const childList = "[".concat(childNames.join(", ")).concat("]");
+            // TODO: revert the console.log. also probably name the child list?
+            stmt = `console.log("reduce", ${childList}.length);\n
+            const ${parName} = ${childList}.reduce((x, y) => x + y); `;
```
Comment (joshpoll) [inline packages/core/src/engine/Autodiff.ts:1494]: fwiw replacing this code with the loop version avoids allocating the closure, thus avoiding the stack blow-up

```suggestion
            let ${parName} = 0;\n
            for (const $child of ${childList}) {\n
              ${parName} += $child;\n
            }`;
```

Comment (hypotext) [reply packages/core/src/engine/Autodiff.ts:1494]: Amazing, thanks! I'll try it out. This is a much simpler solution than the one I thought of originally.

Comment (joshpoll) [reply packages/core/src/engine/Autodiff.ts]: yeah I tried it locally and it seems to work (as long as parName isn't `$child`)

**Comment by wodeni** [comment]
Closing due to inactivity.

**Pull request closed without merging**